openid-zkp-auth/circuits/jwt_proof_ua.circom

```diff
@@ -0,0 +1,170 @@
+    - mask[inCount]:            A binary mask over X, i.e., mask[i] = 0 or 1
+    - randomness:               A 248-bit random number to keep the sensitive parts of JWT hidden
+
+    Circuit signals revealed to the verifier along with the ZK proof:
+    - jwt_sha2_hash:            The SHA2 hash of the JWT header + JWT payload + SHA-2 padding
+    - num_sha2_blocks:          Number of SHA2 (64-byte) blocks the SHA2-padded JWT consumes
+    - subject_id_com:           H(subject_id || PIN). A binding and hiding commitment to subject_id
+    - payload_start_index:      The index of the payload in the content
+    - payload_len:              The length of the payload
+    - masked_content:           The content with the sensitive parts masked
+    - eph_public_key[2]:        The ephemeral public key split into two 128-bit values
+    - max_epoch:                The maximum epoch for which the eph_public_key is valid
+
+    Public Inputs:
+    - all_inputs_hash:          H(jwt_sha2_hash[2] || masked_content_hash || payload_start_index || payload_len
+                                  eph_public_key[2] || max_epoch || nonce || num_sha2_blocks || subject_id_com)
+*/
+template JwtProofUA(maxContentLength, maxSubLength) {
+    // Input is Base64 characters encoded as ASCII
+    var inWidth = 8;
+    var inCount = maxContentLength;
+    signal input content[inCount];
+
+    /**
+     1. SHA2(content)
+    */
+    signal input num_sha2_blocks;
+    var hashCount = 2;
+    var hashWidth = 256 / hashCount;
+    signal jwt_sha2_hash[hashCount] <== Sha2_wrapper(inWidth, inCount, hashWidth, hashCount)(
+        content, 
+        num_sha2_blocks
+    );
+
+    /**
+     2. Checks on subject_id
+        a) Is it in the JWT payload?
+        b) Is subject_id[i] == 0 for all i >= sub_length_ascii?
+        c) Is subject_id[sub_length_ascii - 1] == ',' or '}'?
+    */
+    var subInWidth = 8;
+    signal input subject_id[maxSubLength];
+    signal input sub_length_ascii; // Check if we ensure it is >= 1 and <= maxSubLength
+    signal input sub_length_b64;
+
+    signal input payload_start_index;
+    signal input sub_claim_index_b64;
+    ASCIISubstrExistsInB64(
+        inCount,
+        maxSubLength
+    )(
+        b64Str <== content,
+        BIndex <== sub_claim_index_b64,
+        lenB <== sub_length_b64,
+        A <== subject_id,
+        lenA <== sub_length_ascii,
+        payloadIndex <== payload_start_index
+    );
+
+    // subject_id[i] == 0 for all i >= sub_length_ascii
+    signal sigt[maxSubLength] <== GTBitVector(maxSubLength)(sub_length_ascii);
+    for (var i = 0; i < maxSubLength; i++) {
+        sigt[i] * subject_id[i] === 0;
+    }
+
+    signal lastchar <== SingleMultiplexer(maxSubLength)(subject_id, sub_length_ascii - 1);
+    (lastchar - 44) * (lastchar - 125) === 0; // lastchar = ',' or '}'
+
+    /**
+     3. Calculate commitment to subject_id. We exclude the last character of the subject_id
```

Comment:
why not also the zeros? can't their number change?

Reply:
Not sure what this comment was about. Does it still apply?